src/cache/cache_las.c

```diff
@@ -765,46 +698,123 @@ __wt_las_insert_updates(WT_CURSOR *cursor, WT_BTREE *btree, WT_PAGE *page, WT_MU
+            stop_ts_pair.timestamp = prev_upd->start_ts;
+            stop_ts_pair.txnid = prev_upd->txnid;
+
+            /* Skip TOMBSTONE */
+            if (prev_upd->type == WT_UPDATE_TOMBSTONE) {
+                WT_ASSERT(session, modifies.size > 0);
+                __wt_modify_vector_pop(&modifies, &prev_upd);
+
+                /* The update newer to a TOMBSTONE must be a full update */
+                WT_ASSERT(session, prev_upd->type == WT_UPDATE_STANDARD);
             }
 
-            if (upd->prepare_state == WT_PREPARE_INPROGRESS)
-                ++prepared_insert_cnt;
+            if (prev_upd->type == WT_UPDATE_MODIFY) {
+                WT_ERR(__wt_buf_set(session, prev_full_value, full_value->data, full_value->size));
+                WT_ERR(__wt_modify_apply_item(session, prev_full_value, prev_upd->data, false));
+            } else {
+                WT_ASSERT(session, prev_upd->type == WT_UPDATE_STANDARD);
 
-            WT_ERR(__las_insert_record(session, cursor, btree_id, key, upd, &stop_ts, &stop_txnid));
-            ++insert_cnt;
+                prev_full_value->data = prev_upd->data;
+                prev_full_value->size = prev_upd->size;
+            }
 
             /*
-             * If we encounter subsequent updates with the same timestamp, skip them to avoid
-             * clobbering the entry that we just wrote to lookaside. Only the last thing in the
-             * update list for that timestamp and transaction id matters anyway.
+             * Skip the updates have the same start timestamp and transaction id
+             *
+             * The update older than onpage_upd can be squashed away. Insert a full update anyway to
+             * simplify the code. It will take some extra space but such case should be rare.
              */
-            while (upd->next != NULL && upd->start_ts == upd->next->start_ts &&
-              upd->txnid == upd->next->txnid)
-                upd = upd->next;
-        } while ((upd = upd->next) != NULL);
+            if (upd->start_ts != prev_upd->start_ts || upd->txnid != prev_upd->txnid ||
+              modifies.size == 0) {
+                /*
+                 * Calculate reverse delta. Insert full update for the newest historical record even
+                 * it's a MODIFY.
+                 *
+                 * It is not correct to check prev_upd == list->onpage_upd as we may have aborted
+                 * updates in the middle.
+                 */
+                if (upd->type == WT_UPDATE_MODIFY && modifies.size > 0) {
+                    nentries = MAX_REVERSE_MODIFY_NUM;
+                    if (__wt_calc_modify((WT_SESSION *)session, prev_full_value, full_value,
+                          prev_full_value->size / 10, entries, &nentries) == 0) {
+                        WT_ERR(__wt_modify_pack(cursor, &modify_value, entries, nentries));
+                        WT_ASSERT(session, modify_value != NULL);
```

Comment:
This assert doesn't make much sense to me, `__wt_modify_pack()` can't return without setting `modify_value` AFAICT.

I suggest removing the assert, but changing `__wt_modify_pack()` so that (1) the side-effect return is the last argument it takes, not the second, and that (2) the first thing it does is set `*modifyp = NULL`, those are conventions we follow in the rest of the code, functions that return side-effects should guarantee the side-effect is never left in an unknown state.

Reply:
Changed as suggested.